rust/tableau_summary/src/twb/raw/datasource/substituter.rs

```diff
@@ -0,0 +1,193 @@
+use std::borrow::Cow;
+use tracing::info;
+
+pub struct Substituter<'a, T: ColumnFinder> {
+    pub finder: &'a T,
+}
+
+pub trait ColumnFinder {
+    /// Given the name of a column, find the column's display name.
+    /// If there is no such column, returns None.
+    fn find_column(&self, name: &str) -> Option<Cow<str>>;
+    /// Given the logical source name and the column name, get the display name for the
+    /// column.
+    fn find_column_for_source(&self, source: &str, name: &str) -> Option<Cow<str>>;
+}
+
+impl<'a, T: ColumnFinder> Substituter<'a, T> {
+    /// Given the parameterized string, replace any referenced columns with their Caption'ed representation
+    /// e.g. if the column: `[Calc_12345]` has the caption: `'Orders made'`, then given the string:
+    /// `"CEIL([Calc_12345])"` we will output: `"CEIL([Orders made])"`.
+    /// This will also return a list of dependencies found in the string: at tuple of (datasource, column)
```

Comment:
this again feels a bit like a problem to be solved with a regex

Reply:
if not, at least create more helper functions for each case below, because it's quite a lot to parse at once.

Reply:
Yeah, the problem is 3 fold:
1. We need to do a substitution with extra functionality (need to lookup the column name for substitution). It isn't just substitute "foo" for "bar". This might be doable with the regex crate.
2. The substitution needs to be made in two different use-cases (`[datasource].[column]` or `[column]`). This could be possible, but not without a significant chunk of work to figure out what that regex would look like and how it can work in conjunction with Issue1.
3. Lastly, we are not just substituting, we are also capturing as well. AFAICT, the regex crate doesn't support this.

I've refactored the Substituter to break up the logic better and "hopefully" make it easier to understand.